### PR TITLE
Remove the std::operator== override in CPPUtils.h

### PR DESCRIPTION
**Pull request by nburles**
This removes the operator == override in CPPUtils.h and adds an additional test in ATNDeserializer.cpp.  I haven't found any other cases where the additional testing is required.

I am comparing the results of C++ and Java parsing by calling toStringTree() and doing a direct text comparison... in 10 out of my 3665 inputs, the string output does not match.  However, it's the exact same mismatch with or without this operator== removal so I consider this patch to be safe to apply without detrimental effects.

The differences look like they may just be caused by incorrect handling of 3+ bytes characters in both the C++ and the Java runtimes (possibly just during the string output)... I'll continue to investigate.

### CODE REVIEW

**Comment by nburles** [comment]
Further testing has shown the incorrectness between UTF-8 characters was because of conversion UTF8 - UTF16 (antlrcpp::s2ws) prior to output... removing that means both runtimes act the same way in these cases (still not outputting 3+ byte characters cleanly, but doing the same thing at least!)

There are still 8 inputs that are not working correctly... in these cases, the C++ runtime appears to be picking a node (seemingly at random) and appending all of the following input text to its intended contents (but excepting that node, the entire tree is still created correctly!)  I can move the node to a few other places in the file and still see the behaviour - but I haven't yet worked out any pattern behind it!

Anyway, the remaining issue appears to be completely unrelated to the operator== override so I'll continue with it and create a separate pull request when I've worked out what's going on.

**Comment by mike-lischke** [comment]
That output conversion is only for tests and the way it works does not allow for chars beyond the Unicode BMP. The UTF-16 output doesn't work anyway on Windows, at least I haven't found a way to print the test output properly.

For the == removal: I'm sure there is more work required. We use the equality operation in many places and all have to be checked, if a shared_ptr is involved.

**Comment by nburles** [comment]
I've found that writing the output to a file as UTF-8 works fine... specifically, a sample test would simply be:

```
std::ofstream outputFile("out.txt");
outputFile << "FILE: " << inputFilename << std::endl;
antlr4::ANTLRFileStream input(inputFilename);
MyNamespace::MyLexer lexer(&input);
antlr4::CommonTokenStream tokens(&lexer);
MyNamespace::MyParser parser(&tokens);
std::shared_ptr<antlr4::tree::ParseTree> tree = parser.start_rule();
outputFile << "Parse Tree: " << tree->toStringTree(&parser) << std::endl;
outputFile.close();
return 0;
```

This output is then directly comparable to other ANTLR runtimes, platforms, etc.  Similarly of course outputting to std::cout works also, but only if it avoids Windows' terminal (e.g. output directly piped by Python's subprocess module).

-----

I would also expect more checks to be required, however having removed the == override, the performance hasn't changed (in VS or GCC) and the parse trees generated given my grammar and inputs are still perfectly correct.

I guess others should test other grammars - preferably comparing the parse tree against the Java ANTLR runtime.

File: runtime/Cpp/runtime/src/atn/ATNDeserializer.cpp [hunk 1]
```diff
@@ -483,7 +483,7 @@ ATN ATNDeserializer::deserialize(const std::vector<uint16_t>& input) {
             continue;
           }
 
-          if (transition->target == endState) {
+          if (transition->target == endState || *(transition->target) == *endState) {
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/ATNDeserializer.cpp]: I believe this is wrong. Both target and endState are raw pointers. In Java also the == operator is used here which checks for object equality (which is simply a pointer comparison in Java). Object identity is achived by using the equals() function of Java Object (and its overrides).

**Comment by nburles** [comment]
You're right, I hadn't compared against the Java behaviour.

**Pull request merged**